File: leetcode/c++/8_string_to_integer_atoi.cpp
// https://leetcode.cn/problems/string-to-integer-atoi/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int myAtoi(string s) {
        if (s.length() == 0) return 0;
        int i = 0, flag = 1, res = 0;
        while (s[i] == ' ') {
            if (++i == s.length()) return 0;
        }
        if (s[i] == '-') flag = -1;
        if (s[i] == '-' || s[i] == '+') i++;
        for (i; i < s.length(); i++) {
            if (s[i] < '0' || s[i] > '9') break;
            if (res > INT_MAX / 10 || res == INT_MAX / 10 && s[i] > '7')
                return flag == 1? INT_MAX : INT_MIN;
            res = res * 10 + (s[i] - '0');
        }
        return res * flag;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(1)

File: leetcode/golang/8_string_to_integer_atoi.go
func myAtoi(s string) int {
    if len(s) == 0 {
        return 0
    }
    res, i, flag, boundary := 0, 0, 1, (math.MaxInt32 / 10);
    for s[i] == ' ' {
        i++
        if i == len(s) {
            return 0;
        }
    }
    if s[i] == '-' {
        flag = -1
    }
    if s[i] == '-' || s[i] == '+' {
        i++
    }
    for ; i < len(s); i++ {
        if s[i] < '0' || s[i] > '9' {
            break
        }
        if res > boundary || res == boundary && s[i] > '7' {
            if flag == 1 {
                return math.MaxInt32
            }
            return math.MinInt32
        }
        res = res * 10 + int(s[i] - '0')
    }
    return res * flag
}